In [1]:
#Imported Libaries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
#Reading Json File
df = pd.read_json('domainTags2.json')

In [3]:
# Printing the dataframe
df.head()

_id  \
0  {'$oid': '5d09dfd68501d21d7d729bb1'}   
1  {'$oid': '5d09dfd78501d21d7d729bbb'}   
2  {'$oid': '5d09dfd78501d21d7d729bbc'}   
3  {'$oid': '5d09dfd78501d21d7d729bbe'}   
4  {'$oid': '5d09dfd78501d21d7d729bbf'}   

                                             tags               companyName  \
0                            [finance, insurance]  [21st century insurance]   
1                                 [3d technology]             [3d networks]   
2  [3d technology, augmented reality, e-commerce]                   [3dphy]   
3          [android, enterprise software, mobile]                     [3lm]   
4  [automotive, electronics, enterprise software]                      [3m]   

   domainName                                     crunchbase_url  \
0        21st  https://www.crunchbase.com/organization/21st-c...   
1  3dnetworks  https://www.crunchbase.com/organization/3d-net...   
2       3dphy  https://www.crunchbase.com/organization/3dphy-...   
3         3lm  https://www.crunchbase.com/organization/3lm?ut...   
4          3m  https://www.crunchbase.com/organization/3m?utm...   

            TYPE                              UpdatedOn  \
0  Product-based  {'$date': '2020-01-18T03:39:34.845Z'}   
1  Product-based  {'$date': '2020-01-20T15:10:33.852Z'}   
2  Product-based  {'$date': '2020-01-16T11:44:30.116Z'}   
3  Product-based  {'$date': '2020-01-21T12:50:14.767Z'}   
4  Product-based  {'$date': '2020-01-17T17:56:53.792Z'}   

                                         companyDesc      companyWebsite  \
0  since 1958, 21st century insurance has been de...        www.21st.com   
1                                                NaN  www.3dnetworks.com   
2  3dphy provides a range of 3d virtual and augme...       www.3dphy.com   
3  3lm has already signed deals to provide its te...         www.3lm.com   
4  3m operates as a diversified technology compan...          www.3m.com   

                employeeNumber                                   foundedDate  \
0  {'max': 10000, 'min': 5001}   {'$date': {'$numberLong': '-378658800000'}}   
1                          NaN         {'$date': '2000-01-01T09:00:00.000Z'}   
2       {'max': 50, 'min': 11}         {'$date': '2014-11-03T09:00:00.000Z'}   
3       {'max': 50, 'min': 11}         {'$date': '2010-07-01T09:00:00.000Z'}   
4     {'max': 10001, 'min': 0}  {'$date': {'$numberLong': '-2145884400000'}}   

                                       domain_groups  \
0  {'Fintech/ Payments': 0.825334256973911, 'Insu...   
1                                                NaN   
2                                                NaN   
3             {'Developer Tools': 0.617930901407716}   
4  {'Hardware': 0.6879067738004571, 'Automotive':...   

                                         fundingList  
0                                                NaN  
1                                                NaN  
2  [{'fundingDate': {'$date': '2014-12-18T09:00:0...  
3  [{'fundingDate': {'$date': '2010-07-01T09:00:0...  
4                                                NaN

In [4]:
# Information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7462 entries, 0 to 7461
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   _id             7462 non-null   object
 1   tags            7397 non-null   object
 2   companyName     7462 non-null   object
 3   domainName      7462 non-null   object
 4   crunchbase_url  7462 non-null   object
 5   TYPE            6097 non-null   object
 6   UpdatedOn       7428 non-null   object
 7   companyDesc     6247 non-null   object
 8   companyWebsite  7336 non-null   object
 9   employeeNumber  6604 non-null   object
 10  foundedDate     6603 non-null   object
 11  domain_groups   4406 non-null   object
 12  fundingList     2140 non-null   object
dtypes: object(13)
memory usage: 758.0+ KB


In [5]:
#Checking the null values in the dataframe
df.isnull().sum()

_id                  0
tags                65
companyName          0
domainName           0
crunchbase_url       0
TYPE              1365
UpdatedOn           34
companyDesc       1215
companyWebsite     126
employeeNumber     858
foundedDate        859
domain_groups     3056
fundingList       5322
dtype: int64

#### Handing Missing Values 

In [6]:
df["companyName"].unique

<bound method Series.unique of 0                     [21st century insurance]
1                                [3d networks]
2                                      [3dphy]
3                                        [3lm]
4                                         [3m]
                         ...                  
7457                 [mind spark technologies]
7458                        [jataayu software]
7459     [informix, informix software india .]
7460                   [carpe diem, carpediem]
7461    [national bureau of economic research]
Name: companyName, Length: 7462, dtype: object>

As it has 7462 unique Comapnies which is equal to number of rows in the dataframe. So we shouldnt remove the any row in the dataframe because each column is necessary.


Creating New Columns

As some of the columns have data which need to be changed into appropriate type, so that it will be useful for the analysis and also create some new columns by extracting the data from the present columns

In [7]:
#Converting the companyName from list to the string type
df["companyName"] = df["companyName"].apply(lambda x:x[0])

In [8]:
#Creating new column from the _id column
df["id"] = 0

In [9]:
for i in range(len(df)):
    df["id"][i] = df["_id"][i].get('$oid')

In [10]:
#Creating new column from the UpdateOn column
df["UpdatedDate"] = 0

In [11]:
for i in range(len(df)):
    if not isinstance(df["UpdatedOn"][i],float):
        df["UpdatedDate"][i] = df["UpdatedOn"][i].get('$date')

In [12]:
from datetime import datetime
for i in range(len(df)):
    if isinstance(df["UpdatedDate"][i],str):
        date_string = df["UpdatedDate"][i]
        datetime_obj = datetime.fromisoformat(date_string[:-1])
        formatted_date_time = datetime_obj.strftime("%Y-%m-%d %H:%M:%S")
        df["UpdatedDate"][i] = formatted_date_time

In [13]:
#Creating max and min employee column from the employeenumber column
df["maxemployee"] = 0
df["minemployee"] = 0

In [14]:
for i in range(len(df)):
    if not isinstance(df["employeeNumber"][i],float):
        df["maxemployee"][i] = df["employeeNumber"][i].get('max')

In [15]:
for i in range(len(df)):
    if not isinstance(df["employeeNumber"][i],float):
        df["minemployee"][i] = df["employeeNumber"][i].get('min')

In [16]:
#Creating the foundDaten column from the foundDate column
df["foundDate"] = 0

In [17]:
for i in range(len(df)):
    if not isinstance(df["foundedDate"][i],float):
        df["foundDate"][i] = df["foundedDate"][i].get('$date')

In [18]:
for i in range(len(df)):
    if isinstance(df["foundDate"][i],str):
        date_string = df["foundDate"][i]
        datetime_obj = datetime.fromisoformat(date_string[:-1])
        formatted_date_time = datetime_obj.strftime("%Y-%m-%d %H:%M:%S")
        df["foundDate"][i] = formatted_date_time

In [19]:
from datetime import datetime,timedelta
for i in range(len(df)):
    if isinstance(df["foundDate"][i],dict):
        
        timestamp_milliseconds = df["foundDate"][i].get('$numberLong')

        unix_epoch = datetime(1950, 1, 1)
        adjusted_datetime = unix_epoch + timedelta(milliseconds=abs(int(timestamp_milliseconds)))
        
        # format datetime object as string
        formatted_date_time = adjusted_datetime.strftime("%Y-%m-%d %H:%M:%S")

        df["foundDate"][i] = formatted_date_time

In [20]:
#Extracting only date from the foundDate column
df["foundDate_ymd"] = 0

In [21]:
for i in range(len(df)):
    if isinstance(df["foundDate"][i],str):
        df["foundDate_ymd"][i] = df["foundDate"][i].split(" ")[0]

In [22]:
#Extracting only date from the UpdateDate column
df["UpdateDate_ymd"] = 0

In [23]:
for i in range(len(df)):
    if isinstance(df["UpdatedDate"][i],str):
        df["UpdateDate_ymd"][i] = df["UpdatedDate"][i].split(" ")[0]

In [24]:
df.head()

_id  \
0  {'$oid': '5d09dfd68501d21d7d729bb1'}   
1  {'$oid': '5d09dfd78501d21d7d729bbb'}   
2  {'$oid': '5d09dfd78501d21d7d729bbc'}   
3  {'$oid': '5d09dfd78501d21d7d729bbe'}   
4  {'$oid': '5d09dfd78501d21d7d729bbf'}   

                                             tags             companyName  \
0                            [finance, insurance]  21st century insurance   
1                                 [3d technology]             3d networks   
2  [3d technology, augmented reality, e-commerce]                   3dphy   
3          [android, enterprise software, mobile]                     3lm   
4  [automotive, electronics, enterprise software]                      3m   

   domainName                                     crunchbase_url  \
0        21st  https://www.crunchbase.com/organization/21st-c...   
1  3dnetworks  https://www.crunchbase.com/organization/3d-net...   
2       3dphy  https://www.crunchbase.com/organization/3dphy-...   
3         3lm  https://www.crunchbase.com/organization/3lm?ut...   
4          3m  https://www.crunchbase.com/organization/3m?utm...   

            TYPE                              UpdatedOn  \
0  Product-based  {'$date': '2020-01-18T03:39:34.845Z'}   
1  Product-based  {'$date': '2020-01-20T15:10:33.852Z'}   
2  Product-based  {'$date': '2020-01-16T11:44:30.116Z'}   
3  Product-based  {'$date': '2020-01-21T12:50:14.767Z'}   
4  Product-based  {'$date': '2020-01-17T17:56:53.792Z'}   

                                         companyDesc      companyWebsite  \
0  since 1958, 21st century insurance has been de...        www.21st.com   
1                                                NaN  www.3dnetworks.com   
2  3dphy provides a range of 3d virtual and augme...       www.3dphy.com   
3  3lm has already signed deals to provide its te...         www.3lm.com   
4  3m operates as a diversified technology compan...          www.3m.com   

                employeeNumber                                   foundedDate  \
0  {'max': 10000, 'min': 5001}   {'$date': {'$numberLong': '-378658800000'}}   
1                          NaN         {'$date': '2000-01-01T09:00:00.000Z'}   
2       {'max': 50, 'min': 11}         {'$date': '2014-11-03T09:00:00.000Z'}   
3       {'max': 50, 'min': 11}         {'$date': '2010-07-01T09:00:00.000Z'}   
4     {'max': 10001, 'min': 0}  {'$date': {'$numberLong': '-2145884400000'}}   

                                       domain_groups  \
0  {'Fintech/ Payments': 0.825334256973911, 'Insu...   
1                                                NaN   
2                                                NaN   
3             {'Developer Tools': 0.617930901407716}   
4  {'Hardware': 0.6879067738004571, 'Automotive':...   

                                         fundingList  \
0                                                NaN   
1                                                NaN   
2  [{'fundingDate': {'$date': '2014-12-18T09:00:0...   
3  [{'fundingDate': {'$date': '2010-07-01T09:00:0...   
4                                                NaN   

                         id          UpdatedDate  maxemployee  minemployee  \
0  5d09dfd68501d21d7d729bb1  2020-01-18 03:39:34        10000         5001   
1  5d09dfd78501d21d7d729bbb  2020-01-20 15:10:33            0            0   
2  5d09dfd78501d21d7d729bbc  2020-01-16 11:44:30           50           11   
3  5d09dfd78501d21d7d729bbe  2020-01-21 12:50:14           50           11   
4  5d09dfd78501d21d7d729bbf  2020-01-17 17:56:53        10001            0   

             foundDate foundDate_ymd UpdateDate_ymd  
0  1961-12-31 15:00:00    1961-12-31     2020-01-18  
1  2000-01-01 09:00:00    2000-01-01     2020-01-20  
2  2014-11-03 09:00:00    2014-11-03     2020-01-16  
3  2010-07-01 09:00:00    2010-07-01     2020-01-21  
4  2017-12-31 15:00:00    2017-12-31     2020-01-17

In [25]:
#Creating the new dataframe by dropping the previous columns i.e from which the data is extarcted
df1 = df.drop(columns = ["_id","UpdatedOn","employeeNumber","foundedDate"])

### 1.Return the Competitors of the company

To find the competitors of the Company,first we will find in which domain the company is present i.e from the tags. Then using those tags traverse all the company domains and print those comapnies


The other way is to find by using the domain_groups but it has more null values. so it cant be used for the analysis

The tags column is used to as the primary column for this, so we will remove the null values in this row and store in the separate dataframe

In [26]:
#Dropping the rows which have null values in the tags columnn
df2 = df1.dropna(subset = ['tags'])
df2

tags  \
0                                  [finance, insurance]   
1                                       [3d technology]   
2        [3d technology, augmented reality, e-commerce]   
3                [android, enterprise software, mobile]   
4        [automotive, electronics, enterprise software]   
...                                                 ...   
7457  [business development, digital marketing, ios,...   
7458  [messaging, mobile, software, telecommunicatio...   
7459                               [internet, software]   
7460  [advertising, brand marketing, e-commerce, mar...   
7461                                    [search engine]   

                               companyName             domainName  \
0                   21st century insurance                   21st   
1                              3d networks             3dnetworks   
2                                    3dphy                  3dphy   
3                                      3lm                    3lm   
4                                       3m                     3m   
...                                    ...                    ...   
7457               mind spark technologies  mindsparktechnologies   
7458                      jataayu software            jataayusoft   
7459                              informix               informix   
7460                            carpe diem             carpe-diem   
7461  national bureau of economic research                   nber   

                                         crunchbase_url           TYPE  \
0     https://www.crunchbase.com/organization/21st-c...  Product-based   
1     https://www.crunchbase.com/organization/3d-net...  Product-based   
2     https://www.crunchbase.com/organization/3dphy-...  Product-based   
3     https://www.crunchbase.com/organization/3lm?ut...  Product-based   
4     https://www.crunchbase.com/organization/3m?utm...  Product-based   
...                                                 ...            ...   
7457  https://www.crunchbase.com/organization/mind-s...            NaN   
7458  https://www.crunchbase.com/organization/jataay...            NaN   
7459   https://www.crunchbase.com/organization/informix            NaN   
7460  https://www.crunchbase.com/organization/carpe-...            NaN   
7461  https://www.crunchbase.com/organization/nation...            NaN   

                                            companyDesc  \
0     since 1958, 21st century insurance has been de...   
1                                                   NaN   
2     3dphy provides a range of 3d virtual and augme...   
3     3lm has already signed deals to provide its te...   
4     3m operates as a diversified technology compan...   
...                                                 ...   
7457  mind spark technologies is one of the world&#0...   
7458  a leading provider of products and solutions f...   
7459  informix provides database management systems ...   
7460  we specialist in website design and developmen...   
7461  the national bureau of economic research is a ...   

                      companyWebsite  \
0                       www.21st.com   
1                 www.3dnetworks.com   
2                      www.3dphy.com   
3                        www.3lm.com   
4                         www.3m.com   
...                              ...   
7457  www.mindsparktechnologies.com/   
7458             www.jataayusoft.com   
7459                www.informix.com   
7460            www.carpe-diem.co.uk   
7461                    www.nber.org   

                                          domain_groups  \
0     {'Fintech/ Payments': 0.825334256973911, 'Insu...   
1                                                   NaN   
2                                                   NaN   
3                {'Developer Tools': 0.617930901407716}   
4     {'Hardware': 0.6879067738004571, 'Automotive':...   
...                                                 ...   
7457                 

In [27]:
#Enter the company for which we need to find the competitors
company_name = input("Enter the company name:" )

Enter the company name:


In [28]:
#Finding the company
Company_row= df1[df1['companyName'].apply(lambda x: company_name in x)]

In [29]:
#Extracting the tags of the comapny
tags_list = Company_row.loc[Company_row.index[0], 'tags']

In [30]:
tags_list

['finance', 'insurance']

In [31]:
Competitors = []

In [32]:
for index,row in df2.iterrows():
    tags = row['tags']
    company = row['companyName']
    for x in tags:
        contains = any(x in item for item in tags_list)
        if contains:
            Competitors.append(company)
            break


In [33]:
print("The Competitors of the Company are:",Competitors)

The Competitors of the Company are: ['21st century insurance', 'aci a verisk company', 'aplazame', 'applied data finance', 'acko', 'ameritas', 'aig property casualty apac region', 'allstate solutions', 'apnaloancom', 'aia australia', 'apttus', '2ndquadrant', 'anthem inc', '36氪36krcom', 'angel broking', 'griffin travel pte singapore asia pacific anz', 'adexa', 'aerojet rocketdyne', 'a blockchain cryptocurrency company', 'american expess', 'ameriprise financial services inc', 'allianz global corporate specialty agcs', 'ally', 'asx', 'bharti', 'boardex', 'capital one', 'axa advisors llc', 'bdo india llp', 'aureus analytics', 'bajaj auto finance', 'bcbsnc', 'charles schwab', 'awpl automated workflow', 'bandhan bank', 'bloomberg lp', 'britannia industries', 'cis by deloitte', 'cit', 'assurant', 'axtria ingenious insights', 'bankbazaar', 'bankerbay', 'chartis', 'bank of ireland', 'axis capital', 'betterment', 'capital financial group inc', 'atma education', 'blackrock', 'cassiopae', 'cgi inf

### 2.Fastest Growth and slow Growth

To find the Fastest and Slowest Growth company in the particular domain wee will create a separate new column i.e Growth from which we will find the fatest and slowest growth companies

The growth is calculated as the percenateg of the number of employees in the company to the number of days. It gives the number of days it took to reach the employee strength

Other way to do this is by using the moneyRaised and investors from the fundingList column. But it has almost 71 percent of null values, which cant be used for the analysis

In [39]:
#Creating the new Column Number of days 
df1["NumberofDays"] = 0

In [40]:
from dateutil import parser
for i in range(len(df1)):
    if isinstance(df1["UpdateDate_ymd"][i],str) and isinstance(df1["foundDate_ymd"][i],str):
        string_date1 = df1["UpdateDate_ymd"][i]
        string_date2 = df1["foundDate_ymd"][i]
        datetime1 = parser.parse(string_date1)
        datetime2 = parser.parse(string_date2)
        df1["NumberofDays"][i] = (datetime1 - datetime2).days
        

In [41]:
#Calculating the mean of the employees from the max and min columns and storing in the new column Mean_employees
df1["Mean_Employees"] = 0

In [42]:
for i in range(len(df)):
    df1["Mean_Employees"][i] = int((df1["maxemployee"][i]+df1["minemployee"][i])//2)

In [44]:
# Calculating the growth of each company
df1["Growth"] = round((df1["Mean_Employees"]/df1["NumberofDays"])*100,2)

But some columns has the number of days as zero, which will produce the infinite growth. So we will remove those rows and create a new dataframe

In [47]:
df3 = df1.drop(df1[df1['Growth'] == np.inf].index).reset_index(drop=True)

In [48]:
df3 = df3.dropna(subset = ['tags'])


In [49]:
#Enter the domain to which we need to find the fatest and slowest growth company
domain = input("Enter the doamin:")

Enter the doamin:internet


In [50]:
domain_tags= df3['tags'].apply(lambda x: domain in x)

# create a new DataFrame with the filtered rows
filtered_df = df3[domain_tags]

# find the company with the highest growth percentage
max_growth = filtered_df.loc[filtered_df['Growth'].idxmax(), 'companyName']

# find the company with the lowest growth percentage
min_growth = filtered_df.loc[filtered_df['Growth'].idxmin(), 'companyName']


In [51]:
#Printing the fastest and slowest Growth Company
print(f"Company with the highest growth: {max_growth}")
print(f"Company with the lowest growth: {min_growth}")

Company with the highest growth: vipkid
Company with the lowest growth: informix
